src/daft-scan/src/scan_task_iters/retrieve_parquet_metadata.rs

```diff
@@ -0,0 +1,274 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::Arc,
+};
+
+use common_error::DaftResult;
+use common_file_formats::ParquetSourceConfig;
+use common_runtime::get_io_runtime;
+use daft_io::{get_io_client, IOConfig, IOStatsContext};
+use daft_parquet::read::read_parquet_metadata;
+use futures::future::try_join_all;
+use itertools::Itertools;
+use parquet2::metadata::FileMetaData;
+
+/// Performs Parquet metadata retrieval for the provided iterator of [`NeedsParquetMetadata`]s
+///
+/// This will transform each [`NeedsParquetMetadata`] into its specified [`NeedsParquetMetadata::WithMetadata`] by
+/// performing batched Parquet metadata retrieval whenever [`NeedsParquetMetadata::should_fetch`] is ``true``.
+pub fn batched_parquet_metadata_retrieval<T>(
```

Comment:
why is T generic? we already have the input and output of this function generic

Reply:
The intention was to allow `batched_parquet_metadata_retrieval` to be generic over its input/output iterator types.

In this PR, `batched_parquet_metadata_retrieval ` can transform `Iterator<ScanTaskSplitDecision> -> Iterator<SplittableScanTaskRef>` by retrieving parquet metadata and performing that transformation.

A follow-on PR I would like to do for limit truncation might do something like `Iterator<ScanTaskLimitDecision> -> Iterator<LimitableScanTaskRef>`, which uses the same `batched_parquet_metadata_retrieval` but in a generic way.

We previously talked about having generic iterators and this was how I interpreted the usage, but feel free to lmk if this is wrong.

Reply:
I agree that this might be premature though. Removed the trait definition to be concrete over the Items in the iterator `Iterator<ScanTaskSplitDecision> -> Iterator<SplittableScanTaskRef>`